tests/unit/s2n_tls13_handshake_state_machine_test.c

```diff
@@ -590,6 +643,47 @@ int main(int argc, char **argv)
         EXPECT_SUCCESS(s2n_connection_free(conn));
     }
 
+    /* Test: s2n_conn_set_tls13_handshake_type does not set FULL_HANDSHAKE if 
+     * a pre-shared key has been chosen. */
+    {
+        struct s2n_connection *conn = s2n_connection_new(S2N_CLIENT);
+        struct s2n_psk *psk = NULL;
+
+        EXPECT_OK(s2n_array_pushback(&conn->psk_params.psk_list, (void**) &psk));
+
+        conn->actual_protocol_version = S2N_TLS13;
+        conn->psk_params.chosen_psk = psk;
+
+        EXPECT_NOT_NULL(conn->psk_params.chosen_psk);
+        EXPECT_OK(s2n_conn_set_tls13_handshake_type(conn));
+
+        EXPECT_FALSE(conn->handshake.handshake_type & FULL_HANDSHAKE);
+
+        conn->psk_params.chosen_psk = NULL;
+
+        EXPECT_OK(s2n_conn_set_tls13_handshake_type(conn));
+        EXPECT_TRUE(conn->handshake.handshake_type & FULL_HANDSHAKE);
+
+        EXPECT_SUCCESS(s2n_connection_free(conn));
+    }
+
+    /* Test: s2n_conn_set_tls13_handshake_type errors if a pre-shared key is 
+     * chosen but client auth is required */
```

Comment:
I'm not sure we're handling this case correctly.

The server setting CLIENT_AUTH because client auth is optional certainly shouldn't lead to an error-- we just shouldn't set CLIENT_AUTH.

We should probably do the same for required client auth / the client side. It's required, but only if it's possible.

Reply:
I get how that makes sense when client authentication is optional. But shouldn't we error if client certificate authentication is required  and we have a psk?

Reply:
I don't think this works if we prioritize whether client auth is enabled over whether a PSK was selected. A server could never enable both client auth and session resumption, because any resumed connection would violate the client auth requirement. And if a connection attempted to use a PSK and then had to fall back to a full handshake, that full handshake could never include client auth.